lib/Sema/TypeCheckPropertyWrapper.cpp

```diff
@@ -81,92 +81,124 @@ static VarDecl *findValueProperty(ASTContext &ctx, NominalTypeDecl *nominal,
 
 /// Determine whether we have a suitable wrapped-value initializer within
 /// a property wrapper type.
-static ConstructorDecl *findInitialValueInit(ASTContext &ctx,
-                                             NominalTypeDecl *nominal,
-                                             VarDecl *valueVar,
-                                             Identifier argumentLabel) {
+static ConstructorDecl *findInitialValueInit(
+    ASTContext &ctx,
+    NominalTypeDecl *nominal,
+    VarDecl *valueVar,
+    Identifier argumentLabel) {
+  // Retrieve the type of the 'value' property.
+  if (!valueVar->hasInterfaceType())
+    ctx.getLazyResolver()->resolveDeclSignature(valueVar);
+  Type valueVarType = valueVar->getValueInterfaceType();
+
+  enum class NonViableReason {
+    Failable,
+    ParameterTypeMismatch,
+    Inaccessible,
+  };
+  SmallVector<std::tuple<ConstructorDecl*, NonViableReason, Type>, 2> nonviable;
   SmallVector<ConstructorDecl *, 2> initialValueInitializers;
-  DeclName initName(ctx, DeclBaseName::createConstructor(), {argumentLabel});
+
   SmallVector<ValueDecl *, 2> decls;
-  nominal->lookupQualified(nominal, initName, NL_QualifiedDefault, decls);
+  nominal->lookupQualified(nominal, DeclBaseName::createConstructor(),
+                           NL_QualifiedDefault, decls);
   for (const auto &decl : decls) {
     auto init = dyn_cast<ConstructorDecl>(decl);
-    if (!init || init->getDeclContext() != nominal)
+    if (!init || init->getDeclContext() != nominal || init->getGenericParams())
       continue;
 
-    initialValueInitializers.push_back(init);
-  }
-
-  switch (initialValueInitializers.size()) {
-  case 0:
-    return nullptr;
+    // Check whether every parameter meets one of the following criteria:
+    //   (1) The parameter has a default argument, or
+    //   (2) The parameter has the given argument label.
+    ParamDecl *wrappedValueParam = nullptr;
+    for (auto param : *init->getParameters()) {
+      // Recognize the first parameter with the requested argument label.
+      if (param->getArgumentName() == argumentLabel && !wrappedValueParam) {
+        wrappedValueParam = param;
+        continue;
+      }
 
-  case 1:
-    break;
+      if (param->getDefaultArgumentKind() != DefaultArgumentKind::None)
```

Comment:
`param->isDefaultArgument()` does this already :)

Reply:
Ah, thanks! I'll fix this in my rebase, since things keep shifting under me ;)